client/components/ProductSingle.js

```diff
@@ -1,37 +1,37 @@
 /* eslint-disable react/destructuring-assignment */
+/* eslint-disable no-shadow */
 import React from 'react';
-import axios from 'axios';
-import { Link } from 'react-router-dom';
+import { connect } from 'react-redux';
+import { fetchProductReviews, fetchProduct } from '../store';
 
-class ProductSingle extends React.Component {
-  constructor() {
-    super();
-    this.state = {
-      product: {},
-      reviews: [],
-    };
-  }
-
-  componentDidMount() {
-    // TODO: what's a more efficient way of doing this (that linter likes)?
-    const { match } = this.props;
-    const { params } = match;
-    const { productId } = params;
+const mapStateToProps = (state, props) => {
+  const { products } = state.product;
+  const { reviews } = state.review;
+  return {
+    products,
+    reviews,
+  };
+};
 
-    axios
-      .get(`api/products/${productId}`)
-      .then(res => res.data)
-      .then(product => this.setState({ product }));
+const mapDispatchToProps = dispatch => {
+  return {
+    fetchProductReviews: id => dispatch(fetchProductReviews(id)),
+    fetchProduct: id => dispatch(fetchProduct(id)),
+  };
+};
 
-    axios
-      .get(`api/reviews/${productId}`)
-      .then(res => res.data)
-      .then(reviews => this.setState({ reviews }));
-  }
+class ProductSingle extends React.Component {
+  componentDidMount = () => {
+    const { productId } = this.props.match.params;
+    const { fetchProduct, fetchProductReviews } = this.props;
+    fetchProductReviews(productId);
+    fetchProduct(productId);
+  };
 
   render() {
-    const { product, reviews } = this.state;
-
+    const { products, reviews } = this.props;
+    if (!products.length) return null;
```

Comment:
See my comments in store about creating a product reducer, along with action type and action creator.

Why would this ever be the case that it is empty?  If it is empty, then should never get to this component since are coming to this route via `<Route exact path="/menu/:productId" component={ProductSingle} />`.  Correct?

So probably don't need this line at all.
```suggestion
    if (!product.length) return null;
```

Reply:
When I go directly to the page, it tries to render once with no data then renders again once it's got the data. I was getting "undefined" errors before adding this if statement.